### PR TITLE
Improve the C++ code quality

### LINKED ISSUE #3944: Need to improve the C++ code quality.
### Describe the Code Quality Issue

There are some bad C++ coding habits in the program, such as not initializing variables, not adding braces when the if and for statements are followed by only one line of code, not initializing variables, not using new in the constructor, but using delete in the destructor.

### Additional Context

Do some refactoring.

### Task list for Issue attackers (only for developers)

- [ ] Identify the specific code file or section with the code quality issue.
- [ ] Investigate the issue and determine the root cause.
- [ ] Research best practices and potential solutions for the identified issue.
- [ ] Refactor the code to improve code quality, following the suggested solution.
- [ ] Ensure the refactored code adheres to the project's coding standards.
- [ ] Test the refactored code to ensure it functions as expected.
- [ ] Update any relevant documentation, if necessary.
- [ ] Submit a pull request with the refactored code and a description of the changes made.

### PR DESCRIPTION
**Pull request by mohanchen**
### Reminder
- [ ] Have you linked an issue with this pull request?
- [ ] Have you added adequate unit tests and/or case tests for your pull request?
- [ ] Have you noticed possible changes of behavior below or in the linked issue?
- [ ] Have you explained the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi? (ignore if not applicable)

### Linked Issue
Fix #3944

### CODE REVIEW

File: source/module_hamilt_lcao/module_gint/gint.cpp [hunk 1]
```diff
@@ -16,36 +16,65 @@
```
Comment (kirk0830) [inline source/module_hamilt_lcao/module_gint/gint.cpp:16]: Suggestion
```
/**
 * On the refactor of ABACUS codes
 * Manually allocate-deallocate memory for objects with new/delete and/or malloc/free is out-dated and risky.
 * A better and more safe choice would be smart pointer. However, before refactoring with smart pointer,
 * developer should be clear about life span and ownership of variables/objects/instances expected to be managed
 * automatically by smart pointer. Const reference would also be a good choice.
* /
```

**Review by jinzx10 — APPROVED** [review]
I remember that deleting a nullptr used to be unspecified in some early standard, but the standards of c++11 and onwards explictly state that it is safe to delete a nullptr, which simply has no effect. Therefore, we probably do not need the "if some_ptr != nullptr" check. This check does not resolve the double-delete issue either, since deleting a pointer usually does not alter its value; it merely deallocates memory attached to it. As a result, a deleted pointer would still pass the if statement. I think nowadays one popular way to handle double-delete is to "make if safe to double-delete" instead of preventing double-delete, i.e., to set a pointer to nullptr immediately after it's deletion.

I notice that a few asserts are added to the code. I remember @dyzheng has some proposal about introducing the "__DEBUG" macro to wrap all assert so that a release-build  would skip all those assertions and might have some some performance gain. I'm not aware the latest status of this proposal and it might be helpful to have it discussed some day.

**Comment by mohanchen** [comment]
Thanks! That's very helpful. 

P.S. For most usage of assert, I consider the removal of assert will only gain negligible improved performance. It would be good to have a chance to discuss this issue.

**Comment by kirk0830** [comment]
Thanks for your contribution, however, presently only in maintainment aspect, I am not satisfied with the naming of some files, classes, functions and variables. Together with their implementation, they strongly block rapid develop of ABACUS. For example, 

filename:
grid_technique.cpp
It cannot provide any useful information when one wants to find a function by just browsing folder `module_gint`. Comparatively other files like `fire.cpp`, `msst.cpp`, `nhchain.cpp` are more clear and illustrative. I also know for some cases the implementation might be "too much" if placed in one single `cpp` file but there would be more reasonable categorization strategy. While for two files `md_base.cpp` and `md_func.cpp`, just judge from their names, again their names will confuse developers who not familar with MD codes but need to change or develop something.

class name:
`Grid_Driver` is confusing, it is actually the problem of not-well-designed functionalities of classes. @jinzx10 and/or I will emphasize in somewhere at sometime in future that "it is crucial for EVERY developer seperate workflow functions (or say driver) and calculation functions apart, and most of time only workflow functions are specific and are really needed to be included in class, as characters of class", if so, there will not be a class implementing methods and then another class with similar prefix to "drive" it.

function name:
@jinzx10 poses good ideas about nomenclature of functions, say `[motion]_[owner]_[object]`, for example, `read_abacus_orb`, `cal_meshball_rho`, although what is `meshball` is not explained in `*.h` file, and there are plenty of functions left without annotations.

variable name:
I have seen many functions with abbr. but without any explanation on them, like `pvpR`. What is worse much is `how_many_atoms`, and `normal` in code pieces:
```c++
	int normal=0;
	this->total_atoms_on_grid = 0;
	int nat_local = 0;
	for(int it=0; it<GlobalC::ucell.ntype; it++)
	{
		for(int ia=0; ia<GlobalC::ucell.atoms[it].na; ia++)
		{
			for(int im=0; im< this->meshball_ncells; im++)
			{
				// bcell[iat]: which bcell iat atom is in.
				// ball[im]: relative position of adjacent bcell.
				normal = index2normal[ this->index_atom[iat] + this->index_ball[im] ];
				if(normal >= nbxyz)
				{
					std::cout << " index_atom=" << index_atom[iat] << std::endl;
					std::cout << " index_ball=" << index_ball[im] << std::endl;
					std::cout << " normal=" << normal << std::endl;
					std::cout << " nbxyz=" << nbxyz << std::endl;
					ModuleBase::WARNING_QUIT("Grid_Technique::init_atoms_on_grid","normal >= nbxyz");
				}
				assert(normal>=0);
				int f = ind_bigcell[normal];
				if(!bigcell_on_processor[normal]) continue;
				++how_many_atoms[f];
				++total_atoms_on_grid;
				this->in_this_processor[iat] = true;
			}
			if(this->in_this_processor[iat]) ++nat_local;
			++iat;
		}
	}
```
I also find there are still some variables left not-zero-initialized in place, like line 305 (previous)/301 (after modification) in `grid_bigcell.cpp`:
```c++

	int ii,jj,kk,in_ext,in_normal;
	for(int i=0; i<this->nxe; i++)
```

In summary:
class design
together with "let naming of files, classes, functions and variables be illustrative as much as possible".

**Review by kirk0830 — APPROVED** [review]

File: source/module_hamilt_lcao/module_gint/grid_technique.cpp [hunk 2]
```diff
@@ -26,17 +26,60 @@ Grid_Technique::Grid_Technique()
...
-	delete[] which_atom;
-	delete[] which_bigcell;
-	delete[] which_unitcell;
-	delete[] bcell_start;
-	delete[] in_this_processor;
-	delete[] trace_lo;
+    if(nlocdimg!=nullptr)
+	{
+		delete[] nlocdimg;
+	}
```
Comment (caic99) [inline source/module_hamilt_lcao/module_gint/grid_technique.cpp:32]: delete a nullptr is OK; there is no need to determine if it is a nullptr before deleting it.

**Comment by caic99** [comment]
> P.S. For most usage of assert, I consider the removal of assert will only gain negligible improved performance. It would be good to have a chance to discuss this issue.

That's correct. I have tested the ABACUS build with NDEBUG flag which disabled all `assert`s, and the running time are almost the same. @jinzx10 @mohanchen FYI

**Comment by caic99** [comment]
@mohanchen In my humble opinion, I would suggest using `clang-format` and `clang-tidy` to improve the code quality, thus there is no need to do the chores manually, while maintaining a unified and standard code format.

**Comment by mohanchen** [comment]
> @mohanchen In my humble opinion, I would suggest using `clang-format` and `clang-tidy` to improve the code quality, thus there is no need to do the chores manually, while maintaining a unified and standard code format.

I agree with you. I met some issues while using clang-format. I will fix it next time
